### ISSUE TITLE
Preserve begin end

### ISSUE THREAD

**Issue by Khady** [ocaml-ppx/ocamlformat#491]
There are some situation where we would like to preserve `begin end` over parenthesis. Well, most of the time actually, but there are places where it is more critical than others. A good example is when `begin end` is used to delimit a big block of code or a lambda:

I read in #9 that it was not possible to make a difference between parenthesis and begin/end. I am wondering if it is still the case.

```ocaml
(* original *)
let () =
  Lwt_main.run begin
    let a = "a" in
    let b = "b" in
    let c = "c" in
    Lwt.return "test"
  end

(* formatted *)
let () =
  Lwt_main.run
    (let a = "a" in
     let b = "b" in
     let c = "c" in
     Lwt.return "test")
```

```ocaml
(* original *)
let () =
  List.iter begin fun v ->
    (* do a lot of things *)
    let a = "a" in
    let b = "b" in
    let c = "c" in
    ()
  end values

(* formated *)

let () =
  List.iter
    (fun v ->
      (* do a lot of things *)
      let a = "a" in
      let b = "b" in
      let c = "c" in
      () )
    values
```

**Comment by jberdine** [comment]
`begin` and `end` are not in the parsetree. Neither are parens. So there is no reliable way to preserve the concrete syntax.

Instead, I think you'd need to come up with an algorithmic criterion for which parens you'd rather have as begin/end. This sort of thing generally takes implementing and looking at what it does on a sizable body of code...

**Comment by ygrek** [comment]
how about the following rule : begin end for multiline block, parentheses for oneline

**Comment by gpetiot** [comment]
Some work has been started in #539, the output could be improved if we keep the parentheses for single-line expressions, but if I recall it spawned weird issues with begin/end blocks.

**Comment by gpetiot** [comment]
Closed with #828
